tests/lava/proc/conv/test_models.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+import unittest
+import numpy as np
+import sys
+
+from lava.magma.core.run_configs import RunConfig
+from lava.magma.core.run_conditions import RunSteps
+from lava.proc.conv.process import Conv
+from lava.proc.conv import utils
+from lava.proc.io.source import RingBuffer as SendProcess
+from lava.proc.io.sink import RingBuffer as ReceiveProcess
+
+
+verbose = True if (('-v' in sys.argv) or ('--verbose' in sys.argv)) else False
+TORCH_IS_AVAILABLE = utils.TORCH_IS_AVAILABLE
+
+
+class ConvRunConfig(RunConfig):
+    """Run configuration selects appropriate Conv ProcessModel based on tag:
+    floating point precision or Loihi bit-accurate fixed point precision"""
+    def __init__(self, custom_sync_domains=None, select_tag='fixed_pt'):
+        super().__init__(custom_sync_domains=custom_sync_domains)
+        self.select_tag = select_tag
+
+    def select(self, proc, proc_models):
+        for pm in proc_models:
+            if self.select_tag in pm.tags:
+                return pm
+        raise AssertionError('No legal ProcessModel found.')
+
+
+def setup_conv():
+    """Sets up random convolution setting."""
+    # conv parameter setup
+    groups = np.random.randint(4) + 1
+    in_channels = (np.random.randint(8) + 1) * groups
+    out_channels = (np.random.randint(8) + 1) * groups
+    kernel_size = np.random.randint([9, 9]) + 1
+    stride = np.random.randint([5, 5]) + 1
+    padding = np.random.randint([5, 5])
+    dilation = np.random.randint([4, 4]) + 1
+    weight_dims = [
+        out_channels,
+        kernel_size[0], kernel_size[1],
+        in_channels // groups
+    ]
+    weight = np.random.randint(256, size=weight_dims) - 128
+
+    # input needs to be a certain size
+    # to make sure the output dimension is never negative
+    input_shape = tuple(
+        (np.random.randint([128, 128]) + kernel_size * dilation).tolist()
+        + [in_channels]
+    )
+    output_shape = utils.output_shape(
+        input_shape, out_channels, kernel_size, stride, padding, dilation
+    )
+
+    conv = Conv(
+        input_shape=input_shape,
+        weight=weight,
+        stride=stride,
+        padding=padding,
+        dilation=dilation,
+        groups=groups,
+    )
+
+    params = {
+        'weight': weight,
+        'kernel_size': kernel_size,
+        'stride': stride,
+        'padding': padding,
+        'dilation': dilation,
+        'groups': groups,
+    }
+
+    return conv, input_shape, output_shape, params
+
+
+class TestConvProcessModels(unittest.TestCase):
+    """Tests for all ProcessModels of Conv"""
+
+    def test_source_sink(self):
+        """Test whatever is being sent form source is received at sink."""
+        num_steps = 10
+        shape = np.random.randint([128, 128, 16]) + 1
+        input = np.random.randint(
+            256,
+            size=(shape).tolist() + [num_steps]
+        ) - 128
```

Comment:
Please do not use random numbers in unit tests as they may introduce non-deterministic behavior. If you want to test ranges of values, do so explicitly. If you just need some data, hard-code it or use randomly generated numbers with a fixed random seed.

Reply:
I would like to use random settings to test them otherwise grid search space is quite large. Typically I use initialize the random seed and print out the seed when there is error so that one can debug it. Here I am printing out the dimension settings instead.

Reply:
What if all tests pass on your system and the test fails in CI/CD, or in production? I'd suggest to rather test all qualitatively different settings (even if it takes long) than to do random tests.

What do other people think? @mgkwill @awintel

Reply:
If you need some randomness, you could still seed you RNG but then have an outer loop running multiple instances of your base test with different random numbers. This should give you broad but still deterministic coverage.

Reply:
Yeah, I am doing 10 random runs of same test in a loop/across time at the moment.